### PR TITLE
refactor: Separate Gutenberg editor implementations

### PR DESCRIPTION
**Pull request by dcalhoun**
Closes https://github.com/Automattic/dotcom-forge/issues/9831. 

Separate the Gutenberg Mobile and GutenbergKit fragments to reduce regression risk when modifying one of the two editors. While the editors do share commonalities, there are also significant differences. We also do not iterate on the editors simultaneously or at the same pace. 

## Testing Instructions

Test both the Gutenberg Mobile and GutenbergKit editors, attempting to find regressions with foundational features—text editing, media uploads, etc. Toggle the "experimental block editor" feature flags to switch between the two editors. 

## Regression Notes

1. Potential unintended areas of impact
    Regressions in either of the block editors. 
3. What I did to test those areas of impact (or what existing automated tests I relied on)
    Manually tested both editors. 
4. What automated tests I added (or what prevented me from doing so)
    None, deemed unnecessary for the experimental editor. 

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>GutenbergKitEditorFragment</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/build.gradle [hunk 1]
```diff
@@ -180,8 +180,8 @@ android {
         buildConfigField "boolean", "ENABLE_SITE_MONITORING", "false"
         buildConfigField "boolean", "SYNC_PUBLISHING", "false"
         buildConfigField "boolean", "ENABLE_IN_APP_UPDATES", "false"
-        buildConfigField "boolean", "ENABLE_NEW_GUTENBERG", "false"
-        buildConfigField "boolean", "ENABLE_NEW_GUTENBERG_THEME_STYLES", "false"
+        buildConfigField "boolean", "ENABLE_GUTENBERG_KIT", "false"
```
Comment (dcalhoun) [inline WordPress/build.gradle:183]: The term "new" is problematic, as it provides little context and can become outdated/incorrect in the future. I replaced it with a more meaningful name.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt [hunk 2]
```diff
@@ -1799,13 +1808,14 @@ class EditPostActivity : LocaleAwareActivity(), EditorFragmentActivity, EditorIm
...
-            if (htmlModeMenuStateOn) PostEditorAnalyticsSession.Editor.HTML
-            else
-            (
-                if (isGutenberg) PostEditorAnalyticsSession.Editor.GUTENBERG
-                else PostEditorAnalyticsSession.Editor.CLASSIC
-            )
+            when {
+                htmlModeMenuStateOn -> PostEditorAnalyticsSession.Editor.HTML
+                isGutenberg -> PostEditorAnalyticsSession.Editor.GUTENBERG
+                isGutenbergKit -> PostEditorAnalyticsSession.Editor.GUTENBERG_KIT
```
Comment (dcalhoun) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:1816]: Introduced a new GutenbergKit-specific analytic indicator to denote the different Gutenberg editors.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt [hunk 3]
```diff
@@ -2391,7 +2401,9 @@ class EditPostActivity : LocaleAwareActivity(), EditorFragmentActivity, EditorIm
         override fun getItem(position: Int): Fragment {
             return when (position) {
                 PAGE_CONTENT -> {
-                    if (showGutenbergEditor) {
+                    if (isGutenbergKitEditor && showGutenbergEditor) {
+                        createGutenbergKitEditorFragment()
```
Comment (dcalhoun) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:2405]: This—along with the related fragment files—is the predominant change proposed. The separation is proposed to reduce risk associated with modifying shared logic between the Gutenberg editors, which are diverging and not progressing forward at the same rate.

File: libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java
Comment (dcalhoun) [inline libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:1]: These changes should, in theory, restore this file to its state prior to https://github.com/wordpress-mobile/WordPress-Android/commit/93c47b1f87e70bc3546bbb2553001a5831013485, where it was heavily modified for the addition of GutenbergKit. This theory can be validated in the form of copying and pasting [the implementation](https://github.com/wordpress-mobile/WordPress-Android/blob/3574ac4b1d8463d3bcd42828121ed70dc9e70098/libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java) from the parent commit. 

I.e., checking out this branch and pasting the previous implementation into this file in your editor should result in Git showing little-to-no modifications[^1] via `git diff`. 

[^1]: There are a few stylistic changes—spacing, ordering, etc.

File: libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergKitEditorFragment.java
Comment (dcalhoun) [inline libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergKitEditorFragment.java:1]: This new GutenbergKit implementation is devoid of all Gutenberg Mobile-specific logic. Hence, the smaller line numbers and several no-ops satisfying the shared interface. Long term, we might consider severing the shared interface.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt
Comment (dcalhoun) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:1]: Long term, we might also consider how this implementation could be simplified, creating clear and useful separation between Aztec, Gutenberg Mobile, and GutenbergKit.

**Review by dcalhoun — COMMENTED** [review]
I consider the failed [SonarQube Cloud result](https://github.com/wordpress-mobile/WordPress-Android/pull/21493#issuecomment-2501019847) irrelevant as previous failed runs originated from crossing a code duplication threshold. However, the point of these changes is to duplicate code in order to reduce unnecessary risk associated with modifying code shared between two very different implementations.

**Comment by nbradbury** [comment]
@dcalhoun This is unrelated to this PR but wanted to mention it. While testing I noticed the tool buttons don't always appear correctly. Here I've enabled both bold and italic, but the italic button is solid black.

buttons

File: libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java
Comment (nbradbury) [reply libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:1]: > checking out this branch and pasting the previous implementation into this file in your editor should result in Git showing little-to-no modifications[1](#user-content-fn-1-a3dfbbc09537feb2245b72bd1cbc9b3a)

Just an FYI that I had to implement `onEditorContentChanged` and `onOpenMediaLibrary` for the previous code to work.

Comment (dcalhoun) [reply libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:1]: Thanks. That is expected given the current version of the shared interface requires those methods.

**Comment by dcalhoun** [comment]
> While testing I noticed the tool buttons don't always appear correctly. Here I've enabled both bold and italic, but the italic button is solid black.

Thanks for the note. This relates to the way mobile web browsers apply hover styles and will be addressed within https://github.com/Automattic/dotcom-forge/issues/8678.

**Review by nbradbury — APPROVED** [review]
Changes look good and I'm glad for the separation :shipit:

**Comment by dcalhoun** [comment]
The above [Sentry alert](https://github.com/wordpress-mobile/WordPress-Android/pull/21493#issuecomment-2506911229) seems misguided, as these changes have not yet been published in an release.

**Pull request merged**